lib/pbench/server/api/resources/generate_key.py

```diff
@@ -0,0 +1,147 @@
+from datetime import datetime, timedelta, timezone
+from http import HTTPStatus
+
+from flask import current_app, jsonify, make_response
+from flask.wrappers import Request, Response
+from flask_restful import abort
+import jwt
+from sqlalchemy.exc import IntegrityError, SQLAlchemyError
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    APIAbort,
+    ApiAuthorizationType,
+    ApiBase,
+    ApiContext,
+    ApiMethod,
+    ApiParams,
+    ApiSchema,
+)
+import pbench.server.auth.auth as Auth
+from pbench.server.database.models.api_key import APIKey
+from pbench.server.database.models.audit import OperationCode
+
+
+def generate_key(time_delta: timedelta, user):
+    user_obj = user.get_json()
+    current_utc = datetime.now(timezone.utc)
+    expiration = current_utc + time_delta
+    payload = {
+        "iat": current_utc,
+        "exp": expiration,
+        "user_id": user_obj["id"],
+        "username": user_obj["username"],
+        "audience": Auth.oidc_client.client_id,
+    }
+    try:
+        generated_api_key = jwt.encode(
+            payload, current_app.secret_key, algorithm=Auth._TOKEN_ALG_INT
+        )
+    except (
+        jwt.InvalidIssuer,
+        jwt.InvalidIssuedAtError,
+        jwt.InvalidAlgorithmError,
+        jwt.PyJWTError,
+    ):
+        current_app.logger.exception(
+            "Could not encode the JWT api_key for user: {}", user
+        )
+        abort(
+            HTTPStatus.INTERNAL_SERVER_ERROR,
+            message="INTERNAL ERROR",
+        )
+    return generated_api_key, current_utc, expiration
+
+
+class GenerateKey(ApiBase):
+    def __init__(self, config: PbenchServerConfig):
+        super().__init__(
+            config,
+            ApiSchema(
+                ApiMethod.POST,
+                OperationCode.CREATE,
+                authorization=ApiAuthorizationType.NONE,
+            ),
```

Comment:
Yes, an API key is a stand-in for an authorization/credential -- the creation of one is a security-related event which should definitely be audited.

And, yes, we should have a `DELETE` method which allows an authorized user (or Admin) to revoke an API key.

And, although practice seems to be split on this, I think we should allow authorized users to query and fetch existing keys.

Reply:
I wonder if adding api_key to audit should be a part of this PR but if it's not too much work then yeah we can add it I guess.

Reply:
It would likely be easier to get it right if we add the auditing at the same time that we add the stuff which is supposed to be audited.

Due to simple evolution, we had to retrofit auditing into our existing code; however, going forward, we should avoid that.  New code should come with tests and any required audit/log calls -- these should not be afterthoughts.